src/transport/SessionManager.h

```diff
@@ -174,8 +174,155 @@ class DLL_EXPORT SessionManager : public TransportMgrDelegate, public FabricTabl
     Optional<SessionHandle> AllocateSession(Transport::SecureSession::Type secureSessionType,
                                             const ScopedNodeId & sessionEvictionHint);
 
+    /**
+     * @brief
+     *  A set of templated helper function that call a provided lambda
+     *  on all sessions in the underlying session table that match the provided
+     *  query criteria.
+     *
+     */
+
+    /**
+     * @brief
+     *
+     * Call the provided lambda on sessions that match the provided ScopedNodeId.
+     *
+     */
+    template <typename Function>
+    void ForEachMatchingSession(const ScopedNodeId & node, Function && function)
+    {
+        mSecureSessions.ForEachSession([&](auto * session) {
+            if (session->GetPeer() == node)
+            {
+                function(session);
+            }
+
+            return Loop::Continue;
+        });
+    }
+
+    /**
+     * @brief
+     *
+     * Call the provided lambda on sessions that match the provided fabric index.
+     *
+     */
+    template <typename Function>
+    void ForEachMatchingSession(FabricIndex fabricIndex, Function && function)
+    {
+        mSecureSessions.ForEachSession([&](auto * session) {
+            if (session->GetFabricIndex() == fabricIndex)
+            {
+                function(session);
+            }
+
+            return Loop::Continue;
+        });
+    }
+
+    /**
+     * @brief
+     *
+     * Call the provided lambda on all sessions that match the logical fabric
+     * associated with the provided ScopedNodeId and target the same logical node.
+     *
+     */
+    template <typename Function>
+    CHIP_ERROR ForEachCollidingSession(const ScopedNodeId & node, Function && function)
+    {
+        Crypto::P256PublicKey targetPubKey;
+
+        auto * targetFabric = mFabricTable->FindFabricWithIndex(node.GetFabricIndex());
+        VerifyOrReturnError(targetFabric != nullptr, CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+        auto err = targetFabric->FetchRootPubkey(targetPubKey);
+        VerifyOrDie(err == CHIP_NO_ERROR);
+
+        mSecureSessions.ForEachSession([&](auto * session) {
+            Crypto::P256PublicKey comparePubKey;
+
+            //
+            // It's entirely possible to either come across a PASE session OR, a CASE session
+            // that has yet to be activated (i.e a CASEServer holding onto a SecureSession object
+            // waiting for a Sigma1 message to arrive). Let's skip those.
+            //
+            if (!session->IsCASESession() || session->GetFabricIndex() == kUndefinedFabricIndex)
+            {
+                return Loop::Continue;
+            }
+
+            auto * compareFabric = mFabricTable->FindFabricWithIndex(session->GetFabricIndex());
+            VerifyOrDie(compareFabric != nullptr);
+
+            err = compareFabric->FetchRootPubkey(comparePubKey);
+            VerifyOrDie(err == CHIP_NO_ERROR);
```

Comment:
These VerifyOrDie could be replaced by returning Loop::Continue and not blowing up...

Reply:
I intentionally added a VerifyOrDie since the inability to retrieve a root public key for a valid FabricInfo instance is a code bug that violates the rule that a valid FabricInfo instance always have a root public key associated with it and should never happen. If it does, we should crash to make it apparent that there is a bug somewhere.